### PR TITLE
DOCSUP-5822: Document the accurateCastOrNull function.

### PR DESCRIPTION
**Pull request by ka1bi4**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Documentation (changelog entry is not required)

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

- Documentation for #16724

Detailed description / Documentation draft:

...

### CODE REVIEW

File: docs/en/sql-reference/functions/type-conversion-functions.md [hunk 1]
```diff
@@ -432,61 +492,85 @@ SELECT toTypeName(CAST(x, 'Nullable(UInt16)')) FROM t_null
 
 ## accurateCast(x, T) {#type_conversion_function-accurate-cast}
 
-Converts ‘x’ to the ‘t’ data type. The differente from cast(x, T) is that accurateCast
-does not allow overflow of numeric types during cast if type value x does not fit
-bounds of type T.
+Converts `x` to the `T` data type. The differente from [cast(x, T)](#type_conversion_function-cast) is that `accurateCast`
+does not allow overflow of numeric types during cast if type value `x` does not fit bounds of type `T`.
```
Comment (olgarev) [inline docs/en/sql-reference/functions/type-conversion-functions.md:496]:
```suggestion
Converts `x` to the `T` data type. The difference from [cast(x, T)](#type_conversion_function-cast) is that `accurateCast`
does not allow overflow of numeric types during cast if type value `x` does not fit bounds of type `T`. If the casted value is not representable in the target type, `accurateCast` throws an exception.
```
```suggestion
Converts `x` to the `T` data type. The differente from [cast(x, T)](#type_conversion_function-cast) is that `accurateCast`
does not allow overflow of numeric types during cast if type value `x` does not fit bounds of type `T`.
```

File: docs/en/sql-reference/functions/type-conversion-functions.md [hunk 2]
```diff
@@ -381,19 +427,23 @@ Result:
 
 ## CAST(x, T) {#type_conversion_function-cast}
 
-Converts ‘x’ to the ‘t’ data type. The syntax CAST(x AS t) is also supported.
+Converts input value `x` to the `T` data type. The syntax `CAST(x AS t)` is also supported.
```
Comment (olgarev) [inline docs/en/sql-reference/functions/type-conversion-functions.md:431]: Note, that if value `x` does not fit the bounds of type `T`, the function will overflow. For example, `CAST(-1, 'UInt8')` will return `255`.

File: docs/en/sql-reference/functions/type-conversion-functions.md [hunk 3]
```diff
@@ -432,61 +492,85 @@ SELECT toTypeName(CAST(x, 'Nullable(UInt16)')) FROM t_null
...
 ```
 
+Query:
+
+``` sql
+SELECT
+    cast(-1, 'UInt8') as uint8,
+    cast(128, 'Int8') as int8,
+    cast('Test', 'FixedString(2)') as fixed_string;
+```
```
Comment (olgarev) [inline docs/en/sql-reference/functions/type-conversion-functions.md:564]:
```suggestion
``` sql
SELECT
    accurateCastOrNull(-1, 'UInt8') as uint8,
    accurateCastOrNull(128, 'Int8') as int8,
    accurateCastOrNull('Test', 'FixedString(2)') as fixed_string;
```
```

File: docs/en/sql-reference/functions/type-conversion-functions.md [hunk 4]
```diff
@@ -432,61 +492,85 @@ SELECT toTypeName(CAST(x, 'Nullable(UInt16)')) FROM t_null
...
 └────────────────────────────────────────────┘
 ```
 
+Query:
+
+``` sql
+SELECT
+    cast(-1, 'UInt8') as uint8,
+    cast(128, 'Int8') as int8,
+    cast('Test', 'FixedString(2)') as fixed_string;
```
Comment (olgarev) [inline docs/en/sql-reference/functions/type-conversion-functions.md:563]:
```suggestion
    accurateCastOrNull(-1, 'UInt8') as uint8,
    accurateCastOrNull(128, 'Int8') as int8,
    accurateCastOrNull('Test', 'FixedString(2)') as fixed_string;
```

File: docs/ru/sql-reference/functions/type-conversion-functions.md [hunk 5]
```diff
@@ -427,6 +487,54 @@ SELECT toTypeName(CAST(x, 'Nullable(UInt16)')) FROM t_null
...
+
+Преобразует входное значение `x` в указанный тип данных `T`. Всегда возвращает тип [Nullable](../../sql-reference/data-types/nullable.md), и возвращает [NULL](../../sql-reference/syntax.md#null-literal), если приведенное значение не может быть представлено в целевом типе.
+
+**Синтаксис**
+
+```sql
+accurateCastOrNull(x, T)
+```
+
+**Parameters**
```
Comment (olgarev) [inline docs/ru/sql-reference/functions/type-conversion-functions.md:500]:
```suggestion
**Параметры**
```

File: docs/en/sql-reference/operators/in.md [hunk 6]
```diff
@@ -13,10 +13,28 @@ SELECT (CounterID, UserID) IN ((34, 123), (101500, 456)) FROM ...
 
 If the left side is a single column that is in the index, and the right side is a set of constants, the system uses the index for processing the query.
 
-Don’t list too many values explicitly (i.e. millions). If a data set is large, put it in a temporary table (for example, see the section “External data for query processing”), then use a subquery.
+Don’t list too many values explicitly (i.e. millions). If a data set is large, put it in a temporary table (for example, see the section [External data for query processing](../../engines/table-engines/special/external-data.md)), then use a subquery.
 
 The right side of the operator can be a set of constant expressions, a set of tuples with constant expressions (shown in the examples above), or the name of a database table or SELECT subquery in brackets.
 
+ClickHouse allows different types inside `IN` subquery. For left hand side it applies conversion to the type of right hand side with [accurateCastOrNull](../functions/type-conversion-functions.md#type_conversion_function-accurate-cast_or_null).
```
Comment (olgarev) [inline docs/en/sql-reference/operators/in.md:20]:
```suggestion
ClickHouse allows different types in the left and right parts of `IN` subquery. In this case it converts the left hand side to the type of the right hand side as if the [accurateCastOrNull](../functions/type-conversion-functions.md#type_conversion_function-accurate-cast_or_null) function is applied.
```

File: docs/en/sql-reference/functions/type-conversion-functions.md [hunk 7]
```diff
@@ -395,61 +497,90 @@ SELECT toTypeName(CAST(x, 'Nullable(UInt16)')) FROM t_null
...
-└───────┴──────┴──────────────┘┘
-```
+**Parameters**
+
+-   `x` — Input value.
+-   `T` — The name of the returned data type.
+
+**Returned value**
+
+-   The value in specified data type `T`.
```
Comment (olgarev) [inline docs/en/sql-reference/functions/type-conversion-functions.md:549]:
```suggestion
-   The value, converted to the specified data type `T`.
```

File: docs/ru/sql-reference/functions/type-conversion-functions.md [hunk 8]
```diff
@@ -377,20 +423,27 @@ SELECT uuid = uuid2;
 
 ## CAST(x, T) {#type_conversion_function-cast}
 
-Преобразует x в тип данных t.
-Поддерживается также синтаксис CAST(x AS t).
+Преобразует вхожное значение `x` в указананный тип данных `T`.
```
Comment (olgarev) [inline docs/ru/sql-reference/functions/type-conversion-functions.md:426]:
```suggestion
Преобразует входное значение `x` в указанный тип данных `T`.
```

File: docs/ru/sql-reference/functions/type-conversion-functions.md [hunk 9]
```diff
@@ -377,20 +423,27 @@ SELECT uuid = uuid2;
...
+
+Поддерживается также синтаксис `CAST(x AS t)`.
 
-Пример:
+Обратите внимание, что если значение `x` не соответствует границам типа `T`, функция переполняется. Например, `CAST(-1, 'UInt8')` возвращает 255.
```
Comment (olgarev) [inline docs/ru/sql-reference/functions/type-conversion-functions.md:430]:
```suggestion
Обратите внимание, что если значение `x` не может быть преобразовано к типу `T`, возникает переполнение. Например, `CAST(-1, 'UInt8')` возвращает 255.
```

File: docs/ru/sql-reference/functions/type-conversion-functions.md [hunk 10]
```diff
@@ -427,6 +490,93 @@ SELECT toTypeName(CAST(x, 'Nullable(UInt16)')) FROM t_null
 
 -   Настройка [cast_keep_nullable](../../operations/settings/settings.md#cast_keep_nullable)
 
+## accurateCast(x, T) {#type_conversion_function-accurate-cast}
+
+Преобразует входное значение `x` в указанный тип данных `T`.
+
+Отличие от [cast(x, T)](#type_conversion_function-cast) в том, что `accurateCast` не допускает переполнения числовых типов, если значение типа `x` не соответствует границам типа `T`. Например, `accurateCast(-1, 'UInt8')` вернет ошибку.
```
Comment (olgarev) [inline docs/ru/sql-reference/functions/type-conversion-functions.md:497]:
```suggestion
В отличие от функции [cast(x, T)](#type_conversion_function-cast), `accurateCast` не допускает переполнения при преобразовании числовых типов. Например, `accurateCast(-1, 'UInt8')` вызовет исключение.
```

File: docs/ru/sql-reference/functions/type-conversion-functions.md [hunk 11]
```diff
@@ -427,6 +490,93 @@ SELECT toTypeName(CAST(x, 'Nullable(UInt16)')) FROM t_null
...
+
+``` text
+Code: 70. DB::Exception: Received from localhost:9000. DB::Exception: Value in column Int8 cannot be safely converted into type UInt8: While processing accurateCast(-1, 'UInt8') AS uint8.
+```
+
+## accurateCastOrNull(x, T) {#type_conversion_function-accurate-cast_or_null}
+
+Преобразует входное значение `x` в указанный тип данных `T`.
+
+Всегда возвращает тип [Nullable](../../sql-reference/data-types/nullable.md), и возвращает [NULL](../../sql-reference/syntax.md#null-literal), если приведенное значение не может быть представлено в целевом типе.
```
Comment (olgarev) [inline docs/ru/sql-reference/functions/type-conversion-functions.md:530]:
```suggestion
Всегда возвращает тип [Nullable](../../sql-reference/data-types/nullable.md). Если исходное значение не может быть преобразовано к целевому типу, возвращает [NULL](../../sql-reference/syntax.md#null-literal).
```

File: docs/ru/sql-reference/operators/in.md [hunk 12]
```diff
@@ -13,10 +13,29 @@ SELECT (CounterID, UserID) IN ((34, 123), (101500, 456)) FROM ...
 
 Если слева стоит один столбец, входящий в индекс, а справа - множество констант, то при выполнении запроса, система воспользуется индексом.
 
-Не перечисляйте слишком большое количество значений (миллионы) явно. Если множество большое - лучше загрузить его во временную таблицу (например, смотрите раздел «Внешние данные для обработки запроса»), и затем воспользоваться подзапросом.
+Не перечисляйте слишком большое количество значений (миллионы) явно. Если множество большое - лучше загрузить его во временную таблицу (например, смотрите раздел [Внешние данные для обработки запроса](../../engines/table-engines/special/external-data.md)), и затем воспользоваться подзапросом.
 
 В качестве правой части оператора может быть множество константных выражений, множество кортежей с константными выражениями (показано в примерах выше), а также имя таблицы или подзапрос SELECT в скобках.
 
+ClickHouse допускает различные типы в левой и правой частях подзапроса `IN`. 
+В этом случае он преобразует левую сторону в тип правой стороны, применяя функцию [accurateCastOrNull](../functions/type-conversion-functions.md#type_conversion_function-accurate-cast_or_null).
```
Comment (olgarev) [inline docs/ru/sql-reference/operators/in.md:21]:
```suggestion
Если типы данных в левой и правой частях подзапроса `IN` различаются, ClickHouse преобразует значение в левой части к типу данных из правой части. Преобразование выполняется по аналогии с функцией [accurateCastOrNull](../functions/type-conversion-functions.md#type_conversion_function-accurate-cast_or_null), т.е. тип данных становится [Nullable](../../sql-reference/data-types/nullable.md), а если преобразование не может быть выполнено, возвращается значение [NULL](../../sql-reference/syntax.md#null-literal).
```

File: docs/en/sql-reference/operators/in.md [hunk 13]
```diff
@@ -13,10 +13,28 @@ SELECT (CounterID, UserID) IN ((34, 123), (101500, 456)) FROM ...
 
 If the left side is a single column that is in the index, and the right side is a set of constants, the system uses the index for processing the query.
 
-Don’t list too many values explicitly (i.e. millions). If a data set is large, put it in a temporary table (for example, see the section “External data for query processing”), then use a subquery.
+Don’t list too many values explicitly (i.e. millions). If a data set is large, put it in a temporary table (for example, see the section [External data for query processing](../../engines/table-engines/special/external-data.md)), then use a subquery.
 
 The right side of the operator can be a set of constant expressions, a set of tuples with constant expressions (shown in the examples above), or the name of a database table or SELECT subquery in brackets.
 
+ClickHouse allows different types in the left and right parts of `IN` subquery. In this case it converts the left hand side to the type of the right hand side as if the [accurateCastOrNull](../functions/type-conversion-functions.md#type_conversion_function-accurate-cast_or_null) function is applied.
```
Comment (olgarev) [inline docs/en/sql-reference/operators/in.md:20]:
```suggestion
ClickHouse allows types to differ in the left and the right parts of `IN` subquery. In this case it converts the left side value to the type of the right side, as if the [accurateCastOrNull](../functions/type-conversion-functions.md#type_conversion_function-accurate-cast_or_null) function is applied. That means, that the data type becomes [Nullable](../../sql-reference/data-types/nullable.md), and if the conversion cannot be performed, it returns  [NULL](../../sql-reference/syntax.md#null-literal).
```

**Comment by ka1bi4** [comment]
Hi,
PR is ready for review.
I updated documentation for accurateCastOrNull function and IN operator.
Moreover, fix some query examples. 

After merging master, I got some changes in reinterpretAs(x, T) — it's not my modifications. Maybe, I should marged it again?

File: docs/ru/sql-reference/functions/type-conversion-functions.md [hunk 14]
```diff
@@ -427,6 +490,93 @@ SELECT toTypeName(CAST(x, 'Nullable(UInt16)')) FROM t_null
...
+``` sql
+SELECT cast(-1, 'UInt8') as uint8; 
+```
+
+Результат:
+
+``` text
+┌─uint8─┐
+│   255 │
+└─────
```
Comment (KochetovNicolai) [inline docs/ru/sql-reference/functions/type-conversion-functions.md:512]: ┘

**Review by KochetovNicolai — APPROVED** [review]

File: docs/en/sql-reference/functions/type-conversion-functions.md [hunk 15]
```diff
@@ -281,82 +319,136 @@ If the string has fewer bytes than N, it is padded with null bytes to the right.
...
 
+Result:
+
 ``` text
 ┌─s──────────┬─s_cut─┐
 │ foo\0bar\0 │ foo   │
 └────────────┴───────┘
 ```
 
-## reinterpretAs(x, T) {#type_conversion_function-cast}
```
Comment (KochetovNicolai) [inline docs/en/sql-reference/functions/type-conversion-functions.md:306]: This variant still works for me. It is strange it was removed.

**Comment by olgarev** [comment]
@alexey-milovidov Hi, this PR was approved but not merged into master. Can we close the ticket?

**Comment by KochetovNicolai** [comment]
Well, there is another one pr about reinterpret #20611
Probably, we should merge it first, and then merge docs changes here.

**Comment by olgarev** [comment]
> Well, there is another one pr about reinterpret #20611
> Probably, we should merge it first, and then merge docs changes here.

@KochetovNicolai Hi! The PR we were waiting for is merged now.

**Comment by KochetovNicolai** [comment]
@olgarev Hi!
Let's resolve conflicts in `type-conversion-functions.md`, and then it may be merged.
Do you need some help with it?

**Comment by ka1bi4** [comment]
@KochetovNicolai Hi!
I resolved the conflict.

**Pull request merged**